\documentclass[10pt,letterpaper]{article}
\usepackage[top=0.85in,left=2.75in,footskip=0.75in]{geometry}

\usepackage{listings}
\usepackage[T1]{fontenc}

\usepackage{amsmath,amssymb}
\usepackage{changepage}
\usepackage[utf8x]{inputenc}
\usepackage{textcomp,marvosym}
\usepackage{cite,nameref,hyperref}
\usepackage{microtype}
\DisableLigatures[f]{encoding = *, family = * }

\usepackage[right]{lineno}
\usepackage{array}
\newcolumntype{+}{!{\vrule width 2pt}}

% create \thickcline for thick horizontal lines of variable length
\newlength\savedwidth
\newcommand\thickcline[1]{%
  \noalign{\global\savedwidth\arrayrulewidth\global\arrayrulewidth 2pt}%
  \cline{#1}%
  \noalign{\vskip\arrayrulewidth}%
  \noalign{\global\arrayrulewidth\savedwidth}%
}

% \thickhline command for thick horizontal lines that span the table
\newcommand\thickhline{\noalign{\global\savedwidth\arrayrulewidth\global\arrayrulewidth 2pt}%
\hline
\noalign{\global\arrayrulewidth\savedwidth}}


% Remove comment for double spacing
%\usepackage{setspace}
%\doublespacing

% Text layout
\raggedright
\setlength{\parindent}{0.5cm}
\textwidth 5.25in
\textheight 8.75in

% Bold the 'Figure #' in the caption and separate it from the title/caption with a period
% Captions will be left justified
\usepackage[aboveskip=1pt,labelfont=bf,labelsep=period,justification=raggedright,singlelinecheck=off]{caption}
\renewcommand{\figurename}{Fig}

% Use the PLoS provided BiBTeX style
\bibliographystyle{plos2015}

% Remove brackets from numbering in List of References
\makeatletter
\renewcommand{\@biblabel}[1]{\quad#1.}
\makeatother



% Header and Footer with logo
\usepackage{lastpage,fancyhdr,graphicx}
\usepackage{epstopdf}
%\pagestyle{myheadings}
\pagestyle{fancy}
\fancyhf{}
%\setlength{\headheight}{27.023pt}
%\lhead{\includegraphics[width=2.0in]{PLOS-submission.eps}}
\rfoot{\thepage/\pageref{LastPage}}
\renewcommand{\headrulewidth}{0pt}
\renewcommand{\footrule}{\hrule height 2pt \vspace{2mm}}
\fancyheadoffset[L]{2.25in}
\fancyfootoffset[L]{2.25in}
\lfoot{\today}

%\titlerunning{Null models for Network Enrichment Analysis}
%\authorrunning{G. S. Jeuken and L. K\"{a}ll}

%\Copyright{Gustavo S. Jeuken and Lukas K\"{a}ll}%mandatory, please use full first names. LIPIcs license is "CC-BY";  http://creativecommons.org/licenses/by/3.0/


\begin{document}
\vspace*{0.2in}

% Title must be 250 characters or less.
\begin{flushleft}
{\Large
\textbf\newline{A simple null model for inferences from network enrichment analysis} % Please use "sentence case" for title and headings (capitalize only the first word in a title (or heading), the first word in a subtitle (or subheading), and any proper nouns).
}
\newline
% Insert author names, affiliations and corresponding author email (do not include titles, positions, or degrees).
\\
Gustavo Jeuken\textsuperscript{1} and
Lukas K\"{a}ll\textsuperscript{1,*}
\\
\bigskip
\textbf{1} Science for Life Laboratory, School of Engineering Sciences in Chemistry, Biotechnology and Health, KTH -- Royal Institute of Technology, Box 1031, 17121 Solna, Sweden
\\
\bigskip

% Use the asterisk to denote corresponding authorship and provide email address in note below.
*lukas.kall@scilifelab.se
\end{flushleft}
% Please keep the abstract below 300 words
\section*{Abstract}
A prevailing technique to infer function from lists of identifications, from molecular biological high-throughput experiments, is over-representation analysis, where the identifications are compared to predefined sets of related genes often referred to as pathways. As at least some pathways are known to be incomplete in their annotation, algorithmic efforts have been made to complement them with information from functional association networks. While the terminology varies in the literature, we will here refer to such methods as Network Enrichment Analysis (NEA). Traditionally, the significance of inferences from NEA has been assigned using a null model constructed from randomizations of the network. Here we instead argue for a null model that more directly relates to the set of genes being studied, and have designed one dynamic programming algorithm that calculates the score distribution of NEA scores that makes it possible to assign unbiased mid $p$~values to inferences. We also implemented a random sampling method, carrying out the same task. We demonstrate that our method obtains a superior statistical calibration as compared to the popular NEA inference engine, BinoX, while also providing statistics that are easier to interpret.



\section*{Introduction}

Over-Representation Analysis (ORA) is commonly used to infer function from sets of analytes such as genes, transcripts, proteins or metabolites\cite{tavazoie1999systematic,khatri2012ten,goeman2007analyzing}. The technique estimates which functional modules, such as complexes or pathways, are overrepresented among a set of identified analytes. One prominent application of the technique is expression analysis, where ORA is regularly used to assess alternation in pathway activity by examining significantly different concentrations of analytes between biological conditions, such as disease state or treatment group.

Most ORA methods are assessing the overlap between the investigated set of analytes, the {\em query set}, and a functional module, the {\em pathway set}, using hypergeometric test or a Fisher's exact test. However, variants such as Gene Set Enrichment Analysis (GSEA)\cite{subramanian2005gene} also includes information on expression levels of the analytes of the query set.

A limiting factor of ORA is the quality of the databases defining pathways. While \url{http://pathguide.org} list 708 databases of pathway definitions\cite{bader2006pathguide}, some critique has been voiced concerning the completeness and rigour of current pathway databases. Hence, efforts have been directed to designing methods that extend the pathway definitions, by functional association networks, like STRING\cite{szklarczyk2014string} and FunCoup\cite{ogris2017funcoup}. Instead of directly examining the overlap between the query and pathway sets, one can evaluate the number of links in the functional association network that connect the query and pathway set\cite{alexeyenko2012network, glaab2012enrichnet, mccormack2013statistical, ogris2016novel, signorelli2016neat}. While the terminology varies in the literature, we will here refer to such methods as {\em Network Enrichment Analysis} (NEA).

The significance of the inferences from NEA is assessed by null models that investigate the number of expected random links from the query gene set to the pathway. Previous efforts have settled for network randomization, where the network definitions are randomly constructed by methods with varying degrees of topological conservation in respect to the original network.

% As the often used network randomization techniques are not completely random, such methods often introduce an undesirable bias in the statistics. For instance, the actual set of random networks is much larger than the ones one use in such test, namely the random networks that retain a certain topological characteristic\cite{newman2001random}.
% Also, the somewhat inaccessible process used to randomize networks produces statistics that are equally hard to interpret, as such statistic in practice test a more complex null model than intended.
Yet, the methods that are used for network randomization are often mentally untractable, as they include a big number of steps that are subject to complex constraints.
As an example, in the link assignment plus second-order conservation method \cite{mccormack2013statistical}, links are assigned at random, but sequentially and at each step they must follow constraints that depend on previous link assignments.
While this results in a network that is both randomized and contains some desired topological features, preserved from the original network, these nested constraints make it very hard to verify that the resulting network is representative. In theory there is a space of all possible networks that could be made up from a certain number of nodes and node degree distribution, however, it is not clear that the link assignment and second-order conservation method samples evenly from that distribution.  And, if the resulting networks are not representative, the rejection of the resulting null hypothesis is uninformative, defeating the purpose of the statistical test.

We argue that the network perturbation methods used in most flavours of NEA is a red herring. Users of NEA do not differ from users of ORA and want to test an association between a query set of genes and pathways.  The typical user has used an experimental or computational method to derive a set of query genes and wants to know if it is a random set, or if it is enriched with respect to a pathway.  In this question, the introduction of a network is not relevant. A network may help to increase the sensitivity of ORA methods, but for practical reasons it is better left out of the null hypothesis. We argue that tests of significance in this type of analysis is naturally implemented as tests of the query set. That is, is the query set a randomly selected set of genes or is there an over-representation of well-connected genes in the query set?

Hence, here we employ a different null model, more directly aimed at modelling the lack of association between query and pathway gene sets, which we formulate as ``there are not more links between the query and pathway gene sets than expected by chance''. We present a dynamic programming algorithm, which we dubbed GeneSetDP, that calculates the exact score distribution of any query of a given size. We also implemented one random sampling algorithm for the same task, that we called GeneSetMC. Both algorithms circumvent the need for parametric score distributions for the calculations of significance.
Combining the GeneSetDP and GeneSetMC with the scoring system of the popular method BinoX, we use simulations to demonstrate that both our algorithms produced unbiased statistics.

\section*{Algorithms}

In network-based gene set analysis one scores a query set of genes, $ \mathcal{Q}=\{g_1 \ldots g_Q\}$, from a genome with genes, $\mathcal{G}=\{g_1 \ldots g_G\}$ (i.e. $\mathcal{Q} \subseteq \mathcal{G}$), and how they relate to a pathway, $\mathcal{P}=\{p_1 \ldots p_P\}$. In NEA the pathway maps to the genome through a network, ${x_{ij}}$, where $x_{ij}=1$ if $g_i$ and $p_j$ are connected, or $x_{ij}=0$ otherwise. The pathways are scored by summing up all connections, i.e. the pathway score can be expressed as,

\begin{equation}
s=\sum_{i=1}^Q\sum_{j=1}^P x_{ij}.
\label{eq:sum_ij}
\end{equation}

We wanted to evaluate our score under the null hypothesis, $H_0$ : ``There are not more links between the query and pathway gene sets than expected by chance''. This formulation translates to calculating the distribution of scores that would occur if the $Q$ query genes are randomly selected from the genome, and see how frequently a score as extreme or more extreme than the current score would appear in that distribution. Hence, we wanted to determine a score distribution, {\em i.e.} the number of ways, $N(s)$, we can pick $Q$ genes and obtain a score, $s$, which we will give two methods for below. Given such a score distribution, we can express a mid $p$~value\cite{lancaster1961significance,hwang2001optimality} for obtaining a score, $s$, as,

\begin{equation}
p(s)=\frac{N(s)/2 +\sum_{s'=s+1}^{S} N(s')}{\sum_{s'=0}^{S} N(s')},
\label{eq:pval}
\end{equation}
where $S$ is the maximal score, $s$, that $Q$ genes can obtain from Eq~(\ref{eq:sum_ij}).

We can reformulate Eq~(\ref{eq:sum_ij}) by defining the number of links, $l_i=\sum_{j=1}^P x_{ij}$, which gives us a score,
\begin{equation}
s=\sum_{i=1}^Q l_i.
\label{eq:sum_i}
\end{equation}
Such a number of links, $l_i$, can be pre-computed for all genes in $\mathcal{G}$. For later purposes we also define a mapping, $k_a=\sum_{\{i:l_i=a\}}1$, giving the number of genes having, $a$, links to the investigated pathway.



\subsection*{Random sampling algorithm}

We first implemented a random sampling algorithm to assess $N(s)$. We randomly selected sets of $Q$ genes from $\mathcal{G}$. These sets scores were calculated with Eq~(\ref{eq:sum_i}). We counted the number of times, $F_B(s)$, a score of $s$ was obtained when sampling $B$ gene sets. We see that $F_B(s)$ will approximately have the same shape as $N(s)$ when selecting a large $B$, and we can hence calculate $p$~values by replacing $N$ with $F_B$ in Eq~(\ref{eq:pval}). We refer to this procedure as GeneSetMC.

\subsection*{Computation of the score distribution}

An alternative approach is to calculate the exact score distribution using dynamic programming. This can be done if we formulate $N(s)$ as a recursion, by observing a partial sum function,  $N_a(s,c)$,
 which expresses the number of ways to select a set of $c$ query genes, each with $\le a$  links to the pathway set and obtain a score of $s$. If we know the distribution of the number of ways to select genes with $\le a-1$ genes,  $N_{a-1}(s,c)$, for all $s$ and $c$,   $N_a(s,c)$ can be determined by investigating the ways to select genes with exactly $a$ links to the investigated pathway.   As there are $k_a \choose b$ ways to select $b$ from $k_a$ elements, there contribution to $N_a(s,c)$ from the selection of  $b$ genes with exactly $a$ links are ${k_a \choose b} N_{a-1}(s-ab,c-b)$. When evaluating all possible values of $b$ we see that
 \begin{equation}
N_a(s,c)=\sum_{b=0}^{k_a}{k_a \choose b} N_{a-1}(s-ab,c-b).
\end{equation}
We also see that $N_a(s,c)=0$ for all $s<0$, $c<0$ or $a<0$, with the exception of $N_{-1}(0,0)=1$.

The final score distribution is given by $N(s)=N_R(s,Q)$, where $R=\max_{i}{l_i}$, i.e the largest number of links to the pathway from any gene in the genome.


\subsection*{Implementation}

\begin{lstlisting}[language=Python, caption={ The central part of the dynamic programing algorithm for finding $N(s)$. The function takes the vector of links per gene, $k_a$, as well as the number of query genes, $Q$ as an input. The function depends on two additional functions {\tt find\_maxscore(k,Q)}, which calculates the maximal score a query of size $Q$ can obtain, and {\tt comb(a,b)}, which calculates $ a \choose b $.}, label=lst:gensetdp, captionpos=t, float, abovecaptionskip=-\medskipamount]
def genesetdp(k,Q):
    max_s = find_maxscore(k,Q)
    N = np.zeros((max_s+1,Q+1))
    N[0,0] = 1

    for a in range(len(k)):
        for s in range(max_s,-1,-1):
            for c in range(Q,-1,-1):
                for b in range(k[a],0,-1): # Stop at b=1
                    if c-b>=0 and s-a*b>=0:
                        N[s,c] += comb(k[a], b) * N[s-a*b,c-b]
    return N[:,q]
\end{lstlisting}

There is a memory efficient implementation. We first note that

\[
N_a(s,c)=\sum_{b=0}^{k_a}{k_a \choose b} N_{a-1}(s-ab,c-b)=\sum_{b=1}^{k_a}{k_a \choose b} N_{a-1}(s-ab,c-b) + N_{a-1}(s,c).
\]


This enables us to calculate $N$ in-place, at least as long as we update the elements in $N_a$ in a reverse order so that we do not alter elements from $N_{a-1}$ still needed to update subsequent elements. That is, we do not have to copy the dynamic programming matrix in each iteration over $a$, instead, we can add $\sum_{b=1}^{k_a}{k_a \choose b} N_{a-1}(s-ab,c-b)$ to previous iterations $N_{a-1}(s,c)$, by nested updates over $s \in \{ S, S-1, \ldots, 0 \}$ and $c \in \{ Q, Q-1, \ldots, 0 \}$. For details see Listing \ref{lst:gensetdp}. We refer to this method as GeneSetDP.

\section*{Methods}

We downloaded network definitions and the BinoX software (on 2018-04-28) for comparisons from \url{https://bitbucket.org/sonnhammergroup/binox}. BinoX was run with the default parameters.

We also downloaded the NEA example given at the BinoX web site. The example files include a pathway definition file that groups $6819$ genes into $289$ human pathways and a network definition file that, after thresholding the links with a score of $0.7$, gave $1244992$ links between those genes.

For the purpose of demonstrating our argument, we selected a representative example in the ``Glycolysis/Gluconeogenesis'' pathway, as it contained a number of genes that coincided with the median number of genes of the pathways in the definition file.

\subsection*{Availability}

The source code of an implementation of GeneSetDP and GeneSetMC, as well as code for generating the plots of this paper, is available from \url{https://github.com/statisticalbiotechnology/genesetdp}.

\section*{Results}

We implemented a Python program that reads network and pathway definition files and scores a query sets against a pathway according to Eq~(\ref{eq:sum_i}), using GeneSetDP and GeneSetMC described in the Algorithm section, that enabled us to assign $p$~values according to Eq~(\ref{eq:pval}). We downloaded pathway and network definitions from the BinoX's website and used the same network threshold as BinoX default ($0.7$).

To illustrate the $p$~value calculation procedure, we plotted the score distributions $N(s)$ for a query size of 10, 20 and 30 genes in Fig \ref{fig:score_dist}, using GeneSetDP.

\begin{figure}[htb]
	\begin{center}
		% \includegraphics[width=.9\textwidth]{figures/score_distribuition_multiple.png}
    \end{center}
  \caption{{\bf Score distribution of a query size of 10, 20 and 30 genes against the ``Glycolysis/Gluconeogenesis'' pathway.} GeneSetDP enables us to calculate the full score distribution for a query of a given size, $Q$, i.e. how many ways one can reach a certain score when adding up the scores for $Q$ genes. A red dot marks the maximum score, $S$, for each case.}
  \label{fig:score_dist}
\end{figure}

\subsection*{Test of Calibration}

In order to test the statistical calibration of our method, we calculated $p$~values for 10000 random picks of gene sets of size $Q$ from the investigated genome using both GeneSetDP and GeneSetMC with $B = 100000$. As our selection is random and following our null hypothesis, we expected the resulting $p$~values to be uniformly distributed. To test this we plotted the $p$~values against their quantile in Fig \ref{fig:calibration} for $10000$ randomly assembled queries from the human genome. As a comparison, we also added a calibration curve, under our null hypothesis, for the popular NEA method BinoX\cite{ogris2016novel}. Here, for all three methods, we are only testing the queries for enrichment, that is, we are only concerned with a higher number of links than expected by random, and this differs from the default BinoX setting which tests for both enrichment and depletion.
% Also of note is the fact that, while we display all methods in the same figure, each represents its own calibration, and that the Spearman's correlation between BinoX and our methods is fairly low ($\rho = 0.49$, $0.55$ and $0.55$ for $Q=10$, $20$ and $30$ respectively)

We note that the calibration of GeneSetDP is slightly conservative, i.e. the calculated $p$ values are larger than expected. Meanwhile, BinoX appears strongly anti-conservative, that is, the $p$ values are lower than expected, even when not taking into account significance by depletion.

\begin{figure}[htb]
  \begin{center}
		  \begin{tabular}[t]{c}
				% \includegraphics[width=.9\textwidth]{figures/calibration.png}
		\end{tabular}
  \end{center}
  \caption{{\bf Calibration of GeneSetDP and GeneSetMC.} We selected 10000 random sets of 10, 20 and 30 genes from the investigated genome definition and calculated their $p$ values with GeneSetMC and GeneSetDP against the ``Glycolysis/Gluconeogenesis'' pathway. To test the $p$~values uniformity we plotted them against their {\em normalized rank}, i.e. each $p$~value's (<rank>-0.5)/<total number of $p$~values>. We also added the calibration curve of BinoX, evaluated on the same random sets. The dashed line shows $y = x$ and the dotted lines $y = 2x$ and $y = 0.5x$, for comparison.}
  \label{fig:calibration}
\end{figure}

To test the consistency of the methods, we also investigated the correspondence between the $p$~values reported by GeneSetDP and GeneSetMC. As can be seen in Fig~\ref{fig:pscatter}, the results are consistent as expected.

\begin{figure}[htb]
  \begin{center}
		  \begin{tabular}[t]{c}
        % (Idea from reviewer comment in email 2018-06-08.)
				% \includegraphics[width=.9\textwidth]{figures/scatter_dp_mc.png}
		\end{tabular}
  \end{center}
  \caption{{\bf Correspondence between GeneSetDP's and GeneSetMC's $p$~values.} We selected 10000 random sets of 10, 20 and 30 genes from the investigated genome definition and calculated their $p$ values with GeneSetMC and GeneSetDP against the ``Glycolysis/Gluconeogenesis'' pathway. We subsequently plotted the obtained $p$~values against each other. Again, the dashed line shows $y = x$ and the dotted lines $y = 2x$ and $y = 0.5x$.}
  \label{fig:pscatter}
\end{figure}



\section*{Discussion}

Here we have implemented two methods, GeneSetDP and GeneSetMC, to calculate unbiased $p$~values for inferences from NEA.
Instead of testing our methods' performance in terms of sensitivity, we here instead chose to test the method in terms of the methods statistical accuracy. We argue that more studies should make a point at demonstrating that their methods are well-calibrated, as it is a prerequisite for measuring performance, most importantly when the ground truth of biological effect is not known.
% With this in mind, we demonstrated that both our methods gave a better calibration than the popular BinoX method, by random sampling of subsets of a genome.

Both our implementations, GeneSetDP and GeneSetMC, are intended for measuring the enrichment of links between the query and the pathway set of genes. Most other NEA methods also report depletion of such links. One could easily modify our code to, instead of only measuring the higher scoring tail, measuring both tails of $N(s)$. However, this might result in a drop in sensitivity, so we have so far not implemented such a mechanism.

We made an explicit definition of the null hypothesis we employed, which centers on the query, $H_0$: ``There are not more links between the query and pathway gene sets than expected by chance''. Previous implementations of NEA in the literature use network randomization methods to determine parameters for various types parametric distributions, by perturbations of the investigated network. In practice, such perturbations are very heuristic in their nature, and hence are difficult to use for calculating accurate statistics. Generally, it is easier to understand null models that relate to user actions.

Furthermore, network randomization methods focus on significance of the links in the network instead of the query set itself. In effect, NEA is an attempt to increase the sensitivity of ORA. The standard user of NEA would typically have identified a set of query genes which she wants to evaluate for over-representation in respect to a pathway database. This goal is the same as for users of ORA. Hence, the null model should be of the same nature for NEA and ORA. GeneSetDP and GeneSetMC are two algorithms that prove that one in NEA can test the uniqueness of the query set in respect to a pathway, just as done in ORA.

\section*{Acknowledgments}

LK was supported by a grant from the Swedish Foundation for Strategic Research (BD15-0043).

\begin{thebibliography}{10}

\bibitem{tavazoie1999systematic}
Tavazoie S, Hughes JD, Campbell MJ, Cho RJ, Church GM.
\newblock Systematic determination of genetic network architecture.
\newblock Nature genetics. 1999;22(3):281.

\bibitem{khatri2012ten}
Khatri P, Sirota M, Butte AJ.
\newblock Ten years of pathway analysis: current approaches and outstanding
  challenges.
\newblock PLoS computational biology. 2012;8(2):e1002375.

\bibitem{goeman2007analyzing}
Goeman JJ, B\"{u}hlmann P.
\newblock Analyzing gene expression data in terms of gene sets: methodological
  issues.
\newblock Bioinformatics. 2007;23(8):980--987.

\bibitem{subramanian2005gene}
Subramanian A, Tamayo P, Mootha VK, Mukherjee S, Ebert BL, Gillette MA, et~al.
\newblock Gene set enrichment analysis: a knowledge-based approach for
  interpreting genome-wide expression profiles.
\newblock Proceedings of the National Academy of Sciences.
  2005;102(43):15545--15550.

\bibitem{bader2006pathguide}
Bader GD, Cary MP, Sander C.
\newblock Pathguide: a pathway resource list.
\newblock Nucleic acids research. 2006;34(suppl\_1):D504--D506.

\bibitem{szklarczyk2014string}
Szklarczyk D, Franceschini A, Wyder S, Forslund K, Heller D, Huerta-Cepas J,
  et~al.
\newblock STRING v10: protein--protein interaction networks, integrated over
  the tree of life.
\newblock Nucleic acids research. 2014;43(D1):D447--D452.

\bibitem{ogris2017funcoup}
Ogris C, Guala D, Kaduk M, Sonnhammer EL.
\newblock FunCoup 4: new species, data, and visualization.
\newblock Nucleic acids research. 2017;46(D1):D601--D607.

\bibitem{alexeyenko2012network}
Alexeyenko A, Lee W, Pernemalm M, Guegan J, Dessen P, Lazar V, et~al.
\newblock Network enrichment analysis: extension of gene-set enrichment
  analysis to gene networks.
\newblock {BMC} bioinformatics. 2012;13(1):226.

\bibitem{glaab2012enrichnet}
Glaab E, Baudot A, Krasnogor N, Schneider R, Valencia A.
\newblock EnrichNet: network-based gene set enrichment analysis.
\newblock Bioinformatics. 2012;28(18):i451--i457.

\bibitem{mccormack2013statistical}
McCormack T, Frings O, Alexeyenko A, Sonnhammer EL.
\newblock Statistical assessment of crosstalk enrichment between gene groups in
  biological networks.
\newblock {PloS} one. 2013;8(1):e54945.

\bibitem{ogris2016novel}
Ogris C, Guala D, Helleday T, Sonnhammer EL.
\newblock A novel method for crosstalk analysis of biological networks:
  improving accuracy of pathway annotation.
\newblock Nucleic acids research. 2016;45(2):e8--e8.

\bibitem{signorelli2016neat}
Signorelli M, Vinciotti V, Wit EC.
\newblock NEAT: an efficient network enrichment analysis test.
\newblock BMC bioinformatics. 2016;17(1):352.

\bibitem{lancaster1961significance}
Lancaster HO.
\newblock Significance tests in discrete distributions.
\newblock Journal of the American Statistical Association.
  1961;56(294):223--234.

\bibitem{hwang2001optimality}
Hwang JG, Yang MC.
\newblock An optimality theory for mid p--values In 2$\times$ 2 contingency
  tables.
\newblock Statistica Sinica. 2001; p. 807--826.

\end{thebibliography}

\end{document}
